Added comment section

brec9824 opened the pull request:
Ready for review @uidaho/lambda 

Comment section is added to project view and includes:
- Only authenticated users can comment
- Editable only by owner
- Delete-able only by owner
- Error messages
- Success message
- Time stamp that updates

Also edited the project view (still pretty rough)

Requires database migration!
Solves issue #69 and #85

brec9824 committed:
Basic comment functionality

Adds the database, adds comments to the project, adds username and creation date. (Currently error in routing)

brec9824 committed:
Added missing route

Added route that didnt get added.

brec9824 committed:
Fixed route error

Comments are now working correctly.

brec9824 committed:
Added controller for comments

Moved comments over into its own controller and updated route.

brec9824 committed:
Added ability to delete comments

User must be the owner to delete their comment. Also added a routes for deleting project comments. Also added eloquent relationships between user and their comments.

brec9824 committed:
Fixed name of comments controller

Name of the comments controller was incorrect but is now fixed.

brec9824 committed:
Added eloquent relationships

Added relationships between users and projects.

brec9824 committed:
Added 403 error page

brec9824 committed:
Added editable comments

Comments are editable by redirecting to a new page to edit it. User must be owner of comment or else they will receive an error.

brec9824 committed:
Corrected page name

Edit comment page's name didnt match the router was routing to but is now fixed.

brec9824 committed:
Corrected page's folder name

Edit comment page's folder name was the old folder but has been updated.

brec9824 committed:
Updated routes and function tweaks

Changed some of the routes because project titles with spaces where causing problems.

brec9824 committed:
Hid edit and delete for non-owner

Hid the edit and delete buttons for comments that dont belong to the current user.

brec9824 committed:
Minor changes to the databases

brec9824 committed:
Added update time of comment

The time stamp on the comment will now show either the time of creation or time of last update.

brec9824 committed:
Added time formatting

Time is now displayed in pacif time and it formatted as Month d, YYYY h:m am/pm

brec9824 committed:
Fixed path

Was pointing to wrong folder.

brec9824 committed:
Moved comments to its own page

Comments section is now being @include into the project view page. Comments can now be inserted anywhere.

brec9824 committed:
Fixed incorrect comment time

Corrected method call to get time for comment.

brec9824 committed:
Renamed route

Renamed the route for editing comments.

brec9824 committed:
Added extra routes

Added routes so if user types incorrect routes they are redirected to project list page.

brec9824 committed:
Fixed route

brec9824 committed:
Testing

brec9824 committed:
Reformat project view

Refromatted the project view page to work better with comments.

brec9824 committed:
Reformat project view

Refromatted the project view page to work better with comments.

Commit:
Changed the project view

Made some changes to how the project view page looked.

rickbau5 commented:
Reviewing now

rickbau5 commented on app/Http/Requests/CreateCommentRequest.php:
```diff
@@ -0,0 +1,41 @@
+     *
+     * @return bool
+     */
+    public function authorize()
```
If you check here if the user is logged in, you wouldn't need the `Auth::guest()` check in `ProjectCommentsController` for creating the comment.

rickbau5 replied:
But then again, you're using the `Session::flash` in the other spot. Speaking of that I didn't even see where the message is displayed

brec9824 replied:
Fixed

rickbau5 commented on app/Http/routes.php:
```diff
@@ -43,6 +43,17 @@
+ *  Project Comments Controller   *
+ *--------------------------------*/
+Route::post('project/{project}/comments', 'ProjectCommentsController@addComment');
+Route::get('project/comments/edit/{projectComment}', 'ProjectCommentsController@editComment');
```
L#50 looks good, but I'm not crazy about these Routes really. The urls don't really seem like they represent the actions that they will be doing.

brec9824 replied:
Done

rickbau5 commented on resources/views/errors/403.blade.php:
```diff
@@ -0,0 +1,17 @@
+@section('head')
+    <title>Forbidden (403) | The Squire Project</title>
+@stop
+
```
Nice :+1:

rickbau5 commented on database/migrations/2016_04_07_203735_create_projects_table.php:
```diff
@@ -15,6 +15,7 @@ public function up()
         Schema::create('projects', function(Blueprint $table) {
             $table->increments('id');
             $table->string('title');
+            //$table->integer('user_id')->unsigned()->index();
```
can be removed

brec9824 replied:
Done

rickbau5 commented on app/Policies/ProjectCommentPolicy.php:
```diff
@@ -0,0 +1,21 @@
+use App\ProjectComment;
+use Log;
+use Illuminate\Auth\Access\HandlesAuthorization;
+class ProjectCommentPolicy
```
This is nice, but why the only policy? What's the benefit of doing this over the checking in Requests or on the controller like you're doing for the other methods?

brec9824 replied:
I edited the policies name so it makes more sense where it is being used. I didn't make another custom request because I would just be reusing that same code or using code I didn't need. I just need to check if the user is the owner of the comment. I think it looks cleaner and is easier to understand then having the check directly in the controller.

rickbau5 replied:
Sounds good

rickbau5 commented on app/Policies/ProjectCommentPolicy.php:
```diff
@@ -0,0 +1,21 @@
+    {
+        return $user->id == $projectComment->user_id;
+    }
+}
```
Just a note, GitHub likes a blank line at the end of files.

brec9824 replied:
Ok

rickbau5 commented on database/migrations/2016_04_13_224615_create_project_comments_table.php:
```diff
@@ -0,0 +1,33 @@
+     */
+    public function up()
+    {
+        Schema::create('projectComments', function (Blueprint $table) {
```
convention is that table names are lower_underscore, but it's not a huge deal

brec9824 replied:
Done

rickbau5 commented on resources/views/project/comments.blade.php:
```diff
@@ -0,0 +1,55 @@
+        <li class="list-group-item">
+
+            <p>{{ $comment->comment_body }}</p>
+            <p class="small"><br>
```
I could just imagine Dr J squinting at this and making a comment about how small it is haha

brec9824 replied:
Done

rickbau5 commented on resources/views/project/editcomment.blade.php:
```diff
@@ -0,0 +1,31 @@
+        <div class="col-md-6 col-md-offset-3">
+            <h4>Edit Comment</h4>
+
+            <form class="form-group" method="POST" action="/project/{{ str_replace(' ', '-', $comment->project->title) }}/{{ $comment->id }}">
```
I've update the project model with a function for this, I'll fix this when it comes time to pr my stuff

rickbau5 commented:
Thoughts:
- Editing comments: I'd prefer in line editing of comments.
- Unit Tests :wink: This would be a great place to start!

Otherwise it looks good to me

rickbau5 commented:
Status @brec9824 ?

brec9824 commented:
@rickbau5 working on it right now

Commit:
Multiple fixes and changes

Changed some comment authentication logic
Changed projectComment table to project_comment
Changed error and flash messages css
Renamed custom policy name to be more relevant
Updated route names
Changed 403 page layout and image (its way better)
Resized comments, buttons, and time stamps

Commit:
Updated factory and seeders

Changed it so when seeding database we are making calls to an outside website for images. Now images are being pulled from a sub folder in images called seeds. Also fixed project comments datatbase file.

brec9824 commented:
@rickbau5 ok ready for review again.

rickbau5 commented on database/seeds/ProjectCommentsTableSeeder.php:
```diff
@@ -17,15 +17,17 @@ public function run()
 
         //Creates project comments, users, and projects
-        /*$faker = Faker\Factory::create();
+        /*$dir = "public/images/seeds/";
```
remove?

brec9824 replied:
This whole commented out section is here in case someone wants to have one comment per question per user. They would then only call this seeder and it would set up all the tables.

rickbau5 replied:
Ah. It could be worth making that a separate seeder/command rather than a commented out section, more user friendly.

brec9824 replied:
Ok will do.

rickbau5 commented:
@brec9824 looks good! Some comments in the seeder, but it can be merged once that gets removed!

Commit:
Split seeder into two

rickbau5 commented:
cool. Resolve the conflicts and then it can be merged

brec9824 commented:
Working on it

Commit:
Updated project view page

Made a few small changes.

Commit:
Changed seeder amount

brec9824 commented:
@rickbau5 done

rickbau5 commented:
Images aren't displaying in their full size?
![image](https://cloud.githubusercontent.com/assets/11299264/14663542/d6dff436-0676-11e6-8fa8-37792300bd38.png)

Commit:
Tweaked css for images on project view

Capped them at a max size and added a border.

brec9824 commented:
@rickbau5 fixed

rickbau5 commented:
cool, merging
